src/BSON/Document.c

```diff
@@ -0,0 +1,355 @@
+/*
+ * Copyright 2014-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_bson_encode.h"
+#include "BSON/Document_arginfo.h"
+#include "BSON/Iterator.h"
+
+zend_class_entry* php_phongo_document_ce;
+
+/* Initialize the object and return whether it was successful. An exception will
+ * be thrown on error. */
+static bool php_phongo_document_init(php_phongo_document_t* intern, bson_t* bson)
+{
+	intern->bson = bson;
+
+	return true;
+}
+
+static HashTable* php_phongo_document_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_document_t* intern;
+	HashTable*             props;
+
+	intern = Z_OBJ_DOCUMENT(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 2);
+
+	if (!intern->bson) {
+		return props;
+	}
+
+	{
+		zval data, length;
+
+		ZVAL_STRINGL(&data, (const char*) bson_get_data(intern->bson), intern->bson->len);
+		zend_hash_str_update(props, "data", sizeof("data") - 1, &data);
+
+		ZVAL_LONG(&length, intern->bson->len);
+		zend_hash_str_update(props, "length", sizeof("length") - 1, &length);
+	}
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_Document)
+
+static PHP_METHOD(MongoDB_BSON_Document, fromBSONString)
+{
+	zval                   zv;
+	php_phongo_document_t* intern = NULL;
+	zend_string*           bson_string;
+	const bson_t*          bson;
+	bson_reader_t*         reader;
+	bool                   eof = false;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_STR(bson_string)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	reader = bson_reader_new_from_data((const uint8_t*) ZSTR_VAL(bson_string), ZSTR_LEN(bson_string));
+	if (!(bson = bson_reader_read(reader, NULL))) {
```

Comment:
Do I understand correctly that this method is just going to read X bytes according to the first four bytes (int32) in the data parameter? Assuming that takes us to the end of the buffer, the function succeeds. Any invalid BSON contained within (e.g. invalid type bytes) would result in runtime exceptions during further access.

For future work, we might want to consider exposing an API for `bson_validate` on the Document and ArrayList classes. Alternatively, that could be an option to enable at construct time (via a `from*` method). For performance reasons, I don't think we'd want to validate by default.

And taking this a step further, is there any reason we'd want a common base class for Document and ArrayList? That could also come down the line only if needed. I suppose if we need to share code that's more easily done behind the scenes in C than through PHP inheritance.

Reply:
Correct. The reader will read a single document from the given data and return that. We use that to initialise a `Document` instance without doing any further work on it. The next call to `bson_reader_read` is a sanity check (taken from `php_phongo_bson_data_to_zval_ex`) to ensure we didn't get more than a single BSON document in the stream.

Reply:
I was thinking of having a common ancestor for a BSON document and BSON array, but decided against it. It might simplify things a bit if we could access the underlying structs the same way (without having to check the object CE before calling the appropriate macro), but this affects only a couple of places in the code. In other instances, users will most likely want to know whether they are getting an array (where they are expected to not pay attention to keys) or a document (where keys are important).

Reply:
Opened [PHPC-2175](https://jira.mongodb.org/browse/PHPC-2175) to track the `bson_validate()` suggestion.